src/XMakeBuildEngine/Evaluation/ItemSpec.cs

```diff
@@ -0,0 +1,261 @@
+                    {
+                        // The expression is not of the form "@(X)". Treat as string
+
+                        //  Code corresponds to EngineFileUtilities.GetFileList
+                        var containsEscapedWildcards = EscapingUtilities.ContainsEscapedWildcards(includeSplitEscaped);
+                        var containsRealWildcards = FileMatcher.HasWildcards(includeSplitEscaped);
+
+                        // '*' is an illegal character to have in a filename.
+                        // todo file-system assumption on legal path characters: https://github.com/Microsoft/msbuild/issues/781
+                        if (containsEscapedWildcards && containsRealWildcards)
+                        {
+
+                            // Just return the original string.
+                            builder.Add(new ValueFragment(includeSplitEscaped, itemSpecLocation.File));
+                        }
+                        else if (!containsEscapedWildcards && containsRealWildcards)
+                        {
+                            // Unescape before handing it to the filesystem.
+                            var filespecUnescaped = EscapingUtilities.UnescapeAll(includeSplitEscaped);
+
+                            builder.Add(new GlobFragment(filespecUnescaped));
+                        }
+                        else
+                        {
+                            // No real wildcards means we just return the original string.  Don't even bother 
+                            // escaping ... it should already be escaped appropriately since it came directly
+                            // from the project file
+
+                            builder.Add(new ValueFragment(includeSplitEscaped, itemSpecLocation.File));
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private ItemExpressionFragment<P, I> ProcessItemExpression(string expression, IElementLocation elementLocation, out bool isItemListExpression)
+        {
+            isItemListExpression = false;
+
+            //  Code corresponds to Expander.ExpandSingleItemVectorExpressionIntoItems
+            if (expression.Length == 0)
+            {
+                return null;
+            }
+
+            var capture = Expander<P, I>.ExpandSingleItemVectorExpressionIntoExpressionCapture(expression, ExpanderOptions.ExpandItems, elementLocation);
+
+            if (capture == null)
+            {
+                return null;
+            }
+
+            isItemListExpression = true;
+
+            return new ItemExpressionFragment<P, I>(capture, expression, this);
+        }
+
+        public IEnumerable<I> FilterItems(IEnumerable<I> items)
+        {
+            return items.Where(i => Fragments.Any(f => f.ItemMatches(i.EvaluatedInclude) > 0));
+        }
+
+        public IEnumerable<ItemFragment> FragmentsMatchingItem(string itemToMatch, out int matches)
```

Comment:
I returned the more complex object (enumerable vs int) because:
- it decreases clutter on the caller's side (they can `var` the enumerable but are forced to explicitly declare the int)
- the actual fragments are more interesting than the number of matches